### PR TITLE
Explore: Create menu for short link button

### LINKED ISSUE #59370: UX: Quick way to copy a link to the Explore page with absolute (UNIX) time instead of relative time
### Background / Why we're doing this

Apologies if this issue has been reported already or if there is a way to achieve what I want already. I tried searching for it but couldn't find anything related.

In the `Explore` page, I want to be able to copy a link to the query / result with absolute time (UNIX time) instead of time relative to me. Copying a link with the default time parameters (now-1h to now) means the link gets stale after some time passes. I have a workflow where I often craft queries in the Explore page and save them somewhere later, but the queries should be for a specific time period only in order to highlight the right results.   

Problematic (default) workflow:
1. Go to the Explore page
2. Create a prometheus query / loki logs query
3. Copy a link to the results / query by going to the browser URL bar and CTRL+C'ing that
4. Share link with the team / paste it in a doc
5. Come back to the link a few months later, realise that the link no longer shows the same results because time was set to from `now-1h` to `now` by default.
6. Figure out what was the intended time (around the time the link was sent) and manually set that yourself.

Workaround workflow:
1. Create the query as in step 1-2.
2. Select the time selector in top right
3. Manually set an absolute time, clicking on the current day and adjusting the hours to be around your current time.
4. Copy the link by CTRL+C'ing the browser URL.

Screenshot 2022-11-28 at 09 51 10

Screenshot 2022-11-28 at 09 51 33

New proposed workflow:
1. Create the query as in step 1-2.
2. Have a small "copy" button on the page which copies the link to the page for you (pastes it into your clipboard contents) with time parameters set to be 'now-1h' to 'now' but 'now' is replaced with absolute UNIX timestamps
**OR** have a button in the Time selector to switch to using "absolute time" for ranges such as last 1 hour, last 5 minutes, etc. so that it's "last 1 hour" at the time the time was evaluated.

### Is there existing research for this?

No research yet

### Problems or tasks

- How to provide a quick way of copying a link to prometheus metrics/loki logs with current time instead of relative time. 

### Deliverables

- A button / behaviour change on the Explore page OR a component in the time selector

### PR DESCRIPTION
**Pull request by gelicia**
**What is this feature?**

One of the biggest draws of Explore is copying URLs allow a user to share what theyre seeing with other people. One thing that was not captured as people might like, however, is the exact time range they are looking at without using a non intuitive keybinding. This adds options to the short link button that allows users to easily copy a link with the exact time theyre looking at, even if they are looking at a relative time. 

For example, if Bob is looking at logs over the last 2 hours, they might want to share those results with Sue, who may not look at the link for several hours in the future. Previously, the link would most easily capture that Bob was looking at the last 2 hours of data, and when Sue clicked it 4 hours later, he would see the last two hours of data from when he clicked it and lose the relevant content.

We still keep the one click short link button since people are used to it. But now there's an optional drop down that allows users to copy a link with exact times, meaning Sue would be looking at the same data Bob was, no matter when he clicked the link.

**Why do we need this feature?**

To let our users share links from Explore that are guaranteed to show what they are looking at.

**Who is this feature for?**

For users who want to share links from explore with absolute times. 

**Which issue(s) does this PR fix?**:

Fixes #59370

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

**Comment by gelicia** [comment]
@diegoadams Wanted to go over a couple options on this. Please ignore the actual content of the menu - a description to a menu item is new to grafana ui so I have to style that yet. The options should both be functional though :)

Option 1: Use the same button/dropdown pattern like with the "Run query" button. 
Pros: Can keep the one click copy link action. The new feature does not interfere with this.
Cons: Takes up more real estate. Needing a button and a drop down button next to it requires things to be more spaced out

Option 2 (in the ephemeral): Use the same button/dropdown pattern as the "New" + button that is on the top/search bar.
Pros: Much more compact. Looks cleaner in my opinion. Discoverability of new feature is built in.
Cons: Makes copying a link a 2 click action where it was 1 before.

File: public/app/core/utils/explore.ts [hunk 1]
```diff
@@ -363,10 +363,14 @@ export function getIntervals(range: TimeRange, lowLimit?: string, resolution?: n
 }
 
 export const copyStringToClipboard = (string: string) => {
-  const el = document.createElement('textarea');
-  el.value = string;
-  document.body.appendChild(el);
-  el.select();
-  document.execCommand('copy');
-  document.body.removeChild(el);
+  if (navigator.clipboard && window.isSecureContext) {
```
Comment (gelicia) [inline public/app/core/utils/explore.ts:366]: I ran into an odd issue with the code here in a menu. It would copy once, but not again without a page refresh. I opted to something more similar to what we have elsewhere

https://github.com/grafana/grafana/blob/main/packages/grafana-ui/src/components/ClipboardButton/ClipboardButton.tsx#L92

The bug where it only copies once from the menu will still be there, but only for users with browsers that don't support navigator.clipboard. Since elsewhere in the codebase we use navigator.clipboard without any fallback, this seems to be a small amount of users and I am not concerned about it.

**Comment by diegoadams** [comment]
Thanks @gelicia , these are great. I was initially leaning towards Option 2 because of its compact design. However, considering the high popularity and convenience of the one-click action, I think Option 1 might be the better choice for now. Of course, this isn't set in stone; we can always monitor its reception and adjust based on user feedback over time.

**Review by Eve832 — COMMENTED** [review]
Documentation update looks good! Not submitting a formal approval because there are code changes in this PR

File: public/app/core/utils/explore.ts (hunk 1 above)
Comment (harisrozajac) [reply public/app/core/utils/explore.ts:366]: I've tested this in Chrome and Firefox and it works as expected

**Review by harisrozajac — APPROVED** [review]
🥇

File: public/app/features/explore/ShortLinkButtonMenu.tsx [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+                  const urlState = getUrlStateFromPaneState(pane);
+                  urlState.range = {
+                    to: pane.range.to.valueOf().toString(),
+                    from: pane.range.from.valueOf().toString(),
+                  };
+                  return JSON.stringify({ [exploreIds[i]]: { ...urlState } });
+                } else {
+                  return undefined;
+                }
+              })
```
Comment (Elfo404) [inline public/app/features/explore/ShortLinkButtonMenu.tsx:53]: you can achieve the same by using `Object.entries`

```suggestion
            const urlStates = Object.entries(panes)
              .map(([exploreId, pane], i) => {
                if (!pane) {
                  return;
                }

                const urlState = getUrlStateFromPaneState(pane);
                urlState.range = {
                  to: pane.range.to.valueOf().toString(),
                  from: pane.range.from.valueOf().toString(),
                };
                return JSON.stringify({ [exploreId]: { ...urlState } });
              })
```

File: public/app/features/explore/ShortLinkButtonMenu.tsx [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+    createAndCopyShortLink(url || global.location.href);
+    reportInteraction('grafana_explore_shortened_link_clicked');
+  };
+
+  const MenuActions = () => {
+    return (
+      <Menu>
+        <Menu.Item
+          key="copy-link"
+          label="Copy shortened URL"
```
Comment (Elfo404) [inline public/app/features/explore/ShortLinkButtonMenu.tsx:29]: maybe we can use the same phrasing as the other button for consistency? (also to avoid confusion if someone wonders what is the difference between a "shortened link" and a "shortened URL")
```suggestion
          label={t('explore.toolbar.copy-shortened-link', 'Copy shortened link')}
```

File: public/app/features/explore/ShortLinkButtonMenu.tsx [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+        narrow={true}
+        onClick={() => onCopyShortLink()}
+        aria-label={t('explore.toolbar.copy-shortened-link', 'Copy shortened link')}
+      />
+      <Dropdown overlay={MenuActions} placement="bottom-end" onVisibleChange={setIsOpen}>
+        <ToolbarButton
+          className={styles.buttonContainer}
+          narrow={true}
+          isOpen={isOpen}
+          aria-label={t('explore.toolbar.copy-shortened-link', 'Copy shortened link')}
```
Comment (Elfo404) [inline public/app/features/explore/ShortLinkButtonMenu.tsx:80]: maybe this should have a different label as it doesn't actually copy the link?

File: public/app/features/explore/ShortLinkButtonMenu.tsx [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+          className={styles.menuItem}
+        />
+      </Menu>
+    );
+  };
+
+  // we need the Toolbar button click to be an action separate from opening/closing the menu
+  return (
+    <Stack gap={0} direction="row" alignItems="baseline" wrap={false}>
+      <ToolbarButton
```
Comment (Elfo404) [inline public/app/features/explore/ShortLinkButtonMenu.tsx:66]: maybe we should wrap this in a `ToolbarButtonRow` so we get nice behaviours automatically when users resize their window:

without:
With `ToolbarButtonRow`:
Screenshot 2023-11-02 at 11 15 11

without `ToolbarButtonRow`:
Screenshot 2023-11-02 at 11 15 02

File: public/app/features/explore/ShortLinkButtonMenu.tsx [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+        <Menu.Item
+          key="copy-link"
+          label="Copy shortened URL"
+          icon="link"
+          onClick={() => onCopyShortLink()}
+          className={styles.menuItem}
+        />
+        <Menu.Item
+          key="copy-link-abs-time"
+          label="Copy absolute time URL"
```
Comment (Elfo404) [inline public/app/features/explore/ShortLinkButtonMenu.tsx:36]: maybe using the same initial phrasing may improve consistency a bit?

also shall we translate this and the description in this button as well?

```suggestion
          label="Copy shortened link with absolute time"
```

**Review by Elfo404 — DISMISSED** [review]

File: public/app/features/explore/ShortLinkButtonMenu.tsx [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+        />
+      </Dropdown>
+    </Stack>
+  );
+}
+
+const getStyles = (theme: GrafanaTheme2) => {
+  return {
+    buttonContainer: css({
+      backgroundColor: theme.colors.background.canvas,
```
[the remark this answers is not in the document]
Comment (gelicia) [reply public/app/features/explore/ShortLinkButtonMenu.tsx:90]: I thought the down carrot and the icon were too far away to look like they belong together. And if they were closer, the click targets too small. I still think this, even for the dashboard buttons, but I will remove the background color for consistency.

File: public/app/features/explore/ShortLinkButtonMenu.tsx [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+    }),
+    menuItem: css({
+      alignItems: 'start',
+      '>:nth-child(2)': {
+        marginLeft: theme.spacing(3),
+      },
+    }),
```
[the remark this answers is not in the document]
Comment (gelicia) [reply public/app/features/explore/ShortLinkButtonMenu.tsx:97]: I was waiting to see if anyone had more generic feedback about the description addition to MenuItem and didn't really get any. I do think the styling without this addition is fine Internal discussion here https://raintank-corp.slack.com/archives/C025WTVRBSN/p1698699099211869?thread_ts=1698688460.390859&cid=C025WTVRBSN

**Comment by gelicia** [comment]
There was feedback from @diegoadams on Slack that the expectation was that the icon would persist as a "last used default". I will persist this in local storage

**Comment by diegoadams** [comment]
Thanks @gelicia , adding comment in here so it doesn't get lost in Slack:

The only thing is that when the user selects a sharing option, the

 corresponding icon option should persist like a "last used default'. this is an example from Figma. More in this [chat](https://raintank-corp.slack.com/archives/C02TXB92G6T/p1696568130978689?thread_ts=1696448824.781799&cid=C02TXB92G6T). We can always do it at a later stage if its tricky,  as what you have is also good.
 
 https://github.com/grafana/grafana/assets/95694434/77efda71-676b-45ed-9e8c-12a94e0b44d4

File: public/app/features/explore/ShortLinkButtonMenu.tsx (hunk 7 above)
[the remark this answers is not in the document]
Comment (gelicia) [reply public/app/features/explore/ShortLinkButtonMenu.tsx]: fixed

File: public/app/features/explore/ShortLinkButtonMenu.tsx (hunk 2 above)
Comment (gelicia) [reply public/app/features/explore/ShortLinkButtonMenu.tsx]: fixed!

File: public/app/features/explore/ShortLinkButtonMenu.tsx (hunk 3 above)
Comment (gelicia) [reply public/app/features/explore/ShortLinkButtonMenu.tsx]: fixed!

File: public/app/features/explore/ShortLinkButtonMenu.tsx (hunk 4 above)
Comment (gelicia) [reply public/app/features/explore/ShortLinkButtonMenu.tsx]: fixed!

File: public/app/features/explore/ShortLinkButtonMenu.tsx (hunk 6 above)
Comment (gelicia) [reply public/app/features/explore/ShortLinkButtonMenu.tsx]: fixed!

**Comment by gelicia** [comment]
> The only thing is that when the user selects a sharing option, the corresponding icon option should persist like a "last used default'. this is an example from Figma. More in this [chat]

fixed!

File: public/app/features/explore/ShortLinkButtonMenu.tsx (hunk 5 above)
Comment (gelicia) [reply public/app/features/explore/ShortLinkButtonMenu.tsx]: Fixed! (this is not in the ephemeral, I missed this until later)

File: public/app/features/explore/ShortLinkButtonMenu.tsx [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+
+import { getUrlStateFromPaneState } from './hooks/useStateSync';
+import { selectPanes } from './state/selectors';
+
+export function ShortLinkButtonMenu() {
+  const panes = useSelector(selectPanes);
+  const [isOpen, setIsOpen] = useState(false);
+  const onCopyShortLink = (persistType?: string, url?: string) => {
+    if (persistType !== undefined) {
+      store.set('grafana.explore.shortLinkShareType', persistType);
```
Comment (ifrost) [inline public/app/features/explore/ShortLinkButtonMenu.tsx:19]: @diegoadams Just to double check. Are we sure using icon is a good idea and it has enough visibility? Clicking on that button is probably something that people may not think about a lot so they may not notice it changed.

@gelicia Tooltip is incorrect after toggling

Screenshot 2023-11-06 at 11 24 50

File: public/app/features/explore/ShortLinkButtonMenu.tsx [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+      icon: 'link',
+      getUrl: () => undefined,
+    },
+    {
+      key: 'copy-link-abs-time',
+      label: t('explore.toolbar.copy-shortened-link-abs-time', 'Copy shortened link with absolute time'),
+      icon: 'clock-nine',
+      description: t(
+        'explore.toolbar.copy-shortened-link-abs-time-desc',
+        'Locks the current time for consistent viewing'
```
Comment (ifrost) [inline public/app/features/explore/ShortLinkButtonMenu.tsx:46]: > Locks...

Could just be me, but I find it confusing what "Locks..." actually mean here. At first sight I thought it means it may change my time range in the Explore.

**Review by ifrost — COMMENTED** [review]
I wasn't able to create a link with absolute time range and split view 🤔 

https://github.com/grafana/grafana/assets/745532/fe791438-a26e-42ac-a19c-223f20db17cf

File: public/app/features/explore/ShortLinkButtonMenu.tsx (hunk 10 above)
Comment (gelicia) [reply public/app/features/explore/ShortLinkButtonMenu.tsx:46]: @diegoadams I didnt have it change the selected time range in explore because it seemed like an unrelated action. Should we choose different verbiage here?

File: public/app/features/explore/hooks/useStateSync/migrators/v1.ts [hunk 11]
```diff
@@ -16,16 +16,33 @@ export interface ExploreURLV1 extends BaseExploreURL {
...
     if (!params || !params.panes || typeof params.panes !== 'string') {
-      return {
+      const defaultObj: ExploreURLV1 = {
         schemaVersion: 1,
         panes: {
           [generateExploreId()]: DEFAULT_STATE,
         },
       };
+      // some URL builders will want to put multiple panes in multiple params. If that's the case here, combine them into one object
+      if (params && params.panes && Array.isArray(params.panes)) {
```
Comment (gelicia) [inline public/app/features/explore/hooks/useStateSync/migrators/v1.ts:28]: @Elfo404 Just wondering what you think about this. The URL builder utility we have builds our panes a little differently than we currently accept them. Since it does return a valid URL, I feel like it's more of something we should support in the parser but I can do something else if youd like.

Basically, the [builder](https://github.com/grafana/grafana/blob/main/packages/grafana-data/src/utils/url.ts#L66) will take multiple values and repeat the parameter instead of combining them. This means we end up with something like `explore?panes={'abc':{}}&panes={'def':{}}` when we want `explore?panes={'abc':{}, 'def':{}}`. 

Should we use a different function or is this change to this parser okay?

**Comment by ifrost** [comment]
> Fixes https://github.com/grafana/grafana/issues/77746

How does it fix it and how to verify it? 🤔

**Comment by gelicia** [comment]
> > Fixes #77746
> 
> How does it fix it and how to verify it? 🤔

Eh yeah, sorry. I thought this would be an easy thing to chuck in here but seems like we don't have an easy way to see this. We will need to discuss this one. Took it off the ticket.

File: public/app/features/explore/ShortLinkButtonMenu.tsx (hunk 9 above)
Comment (diegoadams) [reply public/app/features/explore/ShortLinkButtonMenu.tsx:19]: > @diegoadams Just to double check. Are we sure using icon is a good idea and it has enough visibility? Clicking on that button is probably something that people may not think about a lot so they may not notice it changed.

@ifrost , the default will still be the "copy shortened link" icon, but while the user is in a "session" the behaviour will persist as currently. This is the same pattern for the Run Query button.  Its a slow way to introduce the user to this feature. 

When the user hits refresh it should default back to the "copy shortened link" and icon. @gelicia, sorry I wasn't clear on this. 

Screenshot 2023-11-08 at 22 26 02

File: public/app/features/explore/hooks/useStateSync/migrators/v1.ts (hunk 11 above)
Comment (Elfo404) [reply public/app/features/explore/hooks/useStateSync/migrators/v1.ts:28]: where is this builder used?

the explore URL structure is now well defined, i don't think that just because a function generates URL in a wrong way we should support them, we should probably fix that function(or how it is used) instead.

if we expect the `panes` param to be a keyed object we should probably not pass it as an array in https://github.com/grafana/grafana/pull/77336/files#diff-287b3efbe1ba71d24789e1853d050e5208211141d6ac2f46f5488f1095abdf1dR62, i don't think the builder is doing anything wrong, we are simply generating the "panes" param wrongly.

File: public/app/features/explore/ShortLinkButtonMenu.tsx (hunk 10 above)
Comment (diegoadams) [reply public/app/features/explore/ShortLinkButtonMenu.tsx:46]: > @diegoadams I didnt have it change the selected time range in explore because it seemed like an unrelated action. Should we choose different verbiage here?

Perhaps we could try :

- Preserve Time Range 
- Freeze Current Time
- Fix Time to Current Selectio

Comment (ifrost) [reply public/app/features/explore/ShortLinkButtonMenu.tsx:46]: How about "Changes relative time to absolute in the URL"? or "Uses absolute time in the URL"?

**Comment by diegoadams** [comment]
Hey @gelicia, rethinking this whole experience since @Elfo404 explained about the additional links here: issue #59370 

https://github.com/grafana/grafana/assets/95694434/9abf7834-f150-4f74-9527-3d98879cbaee
Screenshot 2023-11-10 at 12 17 14

This way we keep the same style of text 
Screenshot 2023-11-10 at 12 18 34

**Comment by diegoadams** [comment]
Hey @gelicia, rethinking this whole experience since @Elfo404 explained about the additional links here: issue #59370 

https://github.com/grafana/grafana/assets/95694434/8aefbb68-c494-47e0-bc79-6b994b17ddf9

https://github.com/grafana/grafana/assets/95694434/9abf7834-f150-4f74-9527-3d98879cbaee
Screenshot 2023-11-10 at 12 17 14

This way we keep the same style of text 
Screenshot 2023-11-10 at 12 18 34

**Comment by gelicia** [comment]
@diegoadams I can do this but just for your awareness, https://github.com/grafana/grafana/pull/77808 introduced the description as part of the core grafana ui element so we can use that knowing it should become more common. Personally, I like the description text more than a tooltip icon in a dropdown menu. 

I will implement it as specified here however, unless given other instructions.

File: public/app/features/explore/hooks/useStateSync/migrators/v1.ts (hunk 11 above)
Comment (gelicia) [reply public/app/features/explore/hooks/useStateSync/migrators/v1.ts:28]: @Elfo404 The builder is used here https://github.com/grafana/grafana/pull/77336/files#diff-287b3efbe1ba71d24789e1853d050e5208211141d6ac2f46f5488f1095abdf1dR62

**Comment by diegoadams** [comment]
Thanks @gelicia, yeah I did initially think the tooltip was overkill but I didn't know how else to elaborate on the new ability. Over time we can remove it. 

I also looked at trying a mix of description and group text but it got text heavy. I will add a comment to the 77808 Slack chat to see if we can cater for scenarios like this in the future. 

Screenshot 2023-11-14 at 16 56 51

File: public/app/features/explore/hooks/useStateSync/migrators/v1.ts (hunk 11 above)
Comment (Elfo404) [reply public/app/features/explore/hooks/useStateSync/migrators/v1.ts:28]: according to [our own documentation](https://grafana.com/docs/grafana/latest/explore/#generating-explore-urls-from-external-tools) about explore URL format:

>panes is a url-encoded JSON object of panes, where each key is the pane ID and each value is an object matching the following schema:

i'm not saying that using query parameter in that way is wrong in absolute, but that our own API uses a single object with keyed entries, and instead of adapting the API to accept this specific format because of how we implemented the functionality in this PR, we can simply use the expected format in the first place.

>to store multiple items in an array is relatively common
also to nitpick, we are not storing multiple items of an array, we are storing multiple entries of an object that then we recombine at parse time. it is not wrong, but it just complicates the parsing logic for no real benefit

**Comment by gelicia** [comment]
@diegoadams Another question - your latest design removes icons from the menu items, meaning there is no visual cue to the user that their selected option is related to the icon that will show in the top bar. I think this will be confusing. What do you think?

**Comment by diegoadams** [comment]
Thanks for bringing this up @gelicia , I agree and following our slack chat, I have updated the design to include icons and removed the tooltip 

Screenshot 2023-11-16 at 18 43 26

**Comment by diegoadams** [comment]
Thanks @gelicia . I  agree with you. 

Let's put this on hold. There is a lot of discussion around this on Slack. 

I am currently in talks with UX and I will get back once I have further clarity and direction.
We may involve a UX writer to help with the text.

**Comment by gelicia** [comment]
@diegoadams If it helps your conversations with UX and writers, I have implemented the 4 choice with headers dropdown and built an ephemeral.

**Comment by diegoadams** [comment]
Just to keep everyone in the loop. Current UX conversations are considering adding a tooltip next to the chevron button. 

This will clarify the difference between the primary Sharing Icon action and the additional sharing options revealed by the chevron. We've also redesigned the dropdown to 3 choices to keep things simple, with the main action still tied to the Sharing Icon.

image
image

**Comment by gelicia** [comment]
> Just to keep everyone in the loop. Current UX conversations are considering adding a tooltip next to the chevron button.

I dont want to muddy the waters here but here's some developer hot takes.

 - For any UX decision here, it's critical to consider that a user clicking any of these choices only affects their clipboard. This means they can "play" with any of these options nearly consequence free. To me, this means that we do not need to have as much written out inside of the app itself. The user can learn by doing most easily.
 - I would prefer that every single option be available in the dropdown. This means keeping the four action dropdown. This lets the user always see every option they have available.
 - I like the idea that what they select shows up as the icon in the toolbar. We do not have separate icons to show the URL is shortened or not. I do not think a short/not short URL is super important to the user most of the time. Many times its just needing a link and not caring about the length (although the default should be that it does a short URL). If it ends up being important and they get a short or non-short URL and they want the alternative, they will discover this almost immediately after clicking the button and going back to the area they took the action will show them the alternatives. A little setback is easily remedied.

So in short, my recommendation is four options, two headers, the link and clock icons repeated on the options. The icons are redundant but it keeps the visual link between the dropdown options and the toolbar. I have put this in place and updated the ephemeral. I do think the wording should also be reworked here as well.

**Comment by diegoadams** [comment]
Hey @gelicia, thanks for the hottake. I'm still awaiting additional feedback from some UX team members, but to be honest, it seems like we're running in circles with so many people involved. I feel we've sufficiently understood their perspectives, and I think we're well-equipped to take it from here.

Your implementation of the ephemeral feature is great. My only suggestions are:

- Switching to the "share" icon instead of the "link" icon, ensuring consistency with the starting point icon for Normal URLs.

- Streamlining the phrasing for "Normal URLs."

Other than that, everything looks great.

Screenshot 2023-11-22 at 20 41 49

**Comment by diegoadams** [comment]
@gelicia , thanks this looks good

**Comment by diegoadams** [comment]
Ive created the video here: https://drive.google.com/file/d/162peVfOmR-skn-a0Y0WXxbrovux18Q6s/view

File: public/app/features/explore/ShortLinkButtonMenu.tsx [hunk 12]
```diff
@@ -0,0 +1,151 @@
...
+};
+
+export function ShortLinkButtonMenu() {
+  const panes = useSelector(selectPanes);
+  const [isOpen, setIsOpen] = useState(false);
+  const [lastSelected, setLastSelected] = useState(defaultMode);
+  const onCopyLink = (shorten: boolean, url?: string) => {
+    if (shorten) {
+      createAndCopyShortLink(url || global.location.href);
+      reportInteraction('grafana_explore_shortened_link_clicked');
```
Comment (ifrost) [inline public/app/features/explore/ShortLinkButtonMenu.tsx:43]: Should we also track if the absolute time range was selected? (cc. @diegoadams)

File: public/app/features/explore/ShortLinkButtonMenu.tsx [hunk 13]
```diff
@@ -0,0 +1,151 @@
...
+    } else {
+      copyStringToClipboard(
```
Comment (ifrost) [inline public/app/features/explore/ShortLinkButtonMenu.tsx:45]: It'd be nice if generating URLs was in a util outside of this component.

**Review by ifrost — APPROVED** [review]

File: public/app/features/explore/ShortLinkButtonMenu.tsx (hunk 13 above)
Comment (gelicia) [reply public/app/features/explore/ShortLinkButtonMenu.tsx:45]: The difficult process of generating an absolute URL is abstracted out of this component and I don't wish to abstract this any more than it is already. We can follow up with this if we want.

File: public/app/features/explore/ShortLinkButtonMenu.tsx (hunk 12 above)
Comment (gelicia) [reply public/app/features/explore/ShortLinkButtonMenu.tsx]: I added it as a property to the tracking
